In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import matplotlib.pyplot as plt

In [4]:
url='datasets1/bank-full.csv'

In [5]:
df= pd.read_csv(url)

In [6]:
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

In [7]:
df.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

In [8]:
df.drop(['contact','poutcome'], axis=1, inplace=True)
df.head()

age           job  marital  education default  balance housing loan  day  \
0   58    management  married   tertiary      no     2143     yes   no    5   
1   44    technician   single  secondary      no       29     yes   no    5   
2   33  entrepreneur  married  secondary      no        2     yes  yes    5   
3   47   blue-collar  married    unknown      no     1506     yes   no    5   
4   33       unknown   single    unknown      no        1      no   no    5   

  month  duration  campaign  pdays  previous   y  
0   may       261         1     -1         0  no  
1   may       151         1     -1         0  no  
2   may        76         1     -1         0  no  
3   may        92         1     -1         0  no  
4   may       198         1     -1         0  no

In [9]:
df.head()

age           job  marital  education default  balance housing loan  day  \
0   58    management  married   tertiary      no     2143     yes   no    5   
1   44    technician   single  secondary      no       29     yes   no    5   
2   33  entrepreneur  married  secondary      no        2     yes  yes    5   
3   47   blue-collar  married    unknown      no     1506     yes   no    5   
4   33       unknown   single    unknown      no        1      no   no    5   

  month  duration  campaign  pdays  previous   y  
0   may       261         1     -1         0  no  
1   may       151         1     -1         0  no  
2   may        76         1     -1         0  no  
3   may        92         1     -1         0  no  
4   may       198         1     -1         0  no

In [10]:
df['education'] = df['education'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['education'])

df['job'] = df['job'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['job'])

df['default'] = df['default'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['default'])

df['housing'] = df['housing'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['housing'])

df['loan'] = df['loan'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['loan'])

df['day'] = df['day'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['day'])

df['month'] = df['month'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['month'])

df['duration'] = df['duration'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['duration'])

df['campaign'] = df['campaign'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['campaign'])

df['pdays'] = df['pdays'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['pdays'])

df['previous'] = df['previous'].replace('unknown', np.nan)
df = df.dropna(axis=0, subset=['previous'])



df.head()

age           job  marital  education default  balance housing loan  day  \
0   58    management  married   tertiary      no     2143     yes   no    5   
1   44    technician   single  secondary      no       29     yes   no    5   
2   33  entrepreneur  married  secondary      no        2     yes  yes    5   
5   35    management  married   tertiary      no      231     yes   no    5   
6   28    management   single   tertiary      no      447     yes  yes    5   

  month  duration  campaign  pdays  previous   y  
0   may       261         1     -1         0  no  
1   may       151         1     -1         0  no  
2   may        76         1     -1         0  no  
5   may       139         1     -1         0  no  
6   may       217         1     -1         0  no

In [11]:
df.shape

(43193, 15)

In [12]:
default = {'no': 0,'yes': 1}
df.default = [default[item] for item in df.default]
df.head()

age           job  marital  education  default  balance housing loan  day  \
0   58    management  married   tertiary        0     2143     yes   no    5   
1   44    technician   single  secondary        0       29     yes   no    5   
2   33  entrepreneur  married  secondary        0        2     yes  yes    5   
5   35    management  married   tertiary        0      231     yes   no    5   
6   28    management   single   tertiary        0      447     yes  yes    5   

  month  duration  campaign  pdays  previous   y  
0   may       261         1     -1         0  no  
1   may       151         1     -1         0  no  
2   may        76         1     -1         0  no  
5   may       139         1     -1         0  no  
6   may       217         1     -1         0  no

In [13]:
housing = {'no': 0,'yes': 1}
df.housing = [housing[item] for item in df.housing]
df.head()

age           job  marital  education  default  balance  housing loan  day  \
0   58    management  married   tertiary        0     2143        1   no    5   
1   44    technician   single  secondary        0       29        1   no    5   
2   33  entrepreneur  married  secondary        0        2        1  yes    5   
5   35    management  married   tertiary        0      231        1   no    5   
6   28    management   single   tertiary        0      447        1  yes    5   

  month  duration  campaign  pdays  previous   y  
0   may       261         1     -1         0  no  
1   may       151         1     -1         0  no  
2   may        76         1     -1         0  no  
5   may       139         1     -1         0  no  
6   may       217         1     -1         0  no

In [14]:
loan = {'no': 0,'yes': 1}
df.loan = [loan[item] for item in df.loan]
df.head()

age           job  marital  education  default  balance  housing  loan  \
0   58    management  married   tertiary        0     2143        1     0   
1   44    technician   single  secondary        0       29        1     0   
2   33  entrepreneur  married  secondary        0        2        1     1   
5   35    management  married   tertiary        0      231        1     0   
6   28    management   single   tertiary        0      447        1     1   

   day month  duration  campaign  pdays  previous   y  
0    5   may       261         1     -1         0  no  
1    5   may       151         1     -1         0  no  
2    5   may        76         1     -1         0  no  
5    5   may       139         1     -1         0  no  
6    5   may       217         1     -1         0  no

In [15]:
y = {'no': 0,'yes': 1}
df.y = [y[item] for item in df.y]
df.head()

age           job  marital  education  default  balance  housing  loan  \
0   58    management  married   tertiary        0     2143        1     0   
1   44    technician   single  secondary        0       29        1     0   
2   33  entrepreneur  married  secondary        0        2        1     1   
5   35    management  married   tertiary        0      231        1     0   
6   28    management   single   tertiary        0      447        1     1   

   day month  duration  campaign  pdays  previous  y  
0    5   may       261         1     -1         0  0  
1    5   may       151         1     -1         0  0  
2    5   may        76         1     -1         0  0  
5    5   may       139         1     -1         0  0  
6    5   may       217         1     -1         0  0

In [16]:
# Create a set of dummy variables from the marital variable
df_marital = pd.get_dummies(df['marital'])

# Join the dummy variables to the main dataframe
df = pd.concat([df, df_marital], axis=1)
df
#remove the marital column
df.drop(['marital'], axis=1, inplace=True)

# Create a set of dummy variables from the job variable
df_job = pd.get_dummies(df['job'])

# Join the dummy variables to the main dataframe
df = pd.concat([df, df_job], axis=1)
df
#remove the marital column
df.drop(['job'], axis=1, inplace=True)

# Create a set of dummy variables from the education variable
df_education = pd.get_dummies(df['education'])

# Join the dummy variables to the main dataframe
df = pd.concat([df, df_education], axis=1)
df
#remove the marital column
df.drop(['education'], axis=1, inplace=True)

# Create a set of dummy variables from the month variable
df_month = pd.get_dummies(df['month'])

# Join the dummy variables to the main dataframe
df = pd.concat([df, df_month], axis=1)
df
#remove the month column
df.drop(['month'], axis=1, inplace=True)
df.head()

age  default  balance  housing  loan  day  duration  campaign  pdays  \
0   58        0     2143        1     0    5       261         1     -1   
1   44        0       29        1     0    5       151         1     -1   
2   33        0        2        1     1    5        76         1     -1   
5   35        0      231        1     0    5       139         1     -1   
6   28        0      447        1     1    5       217         1     -1   

   previous  ...  dec  feb  jan  jul  jun  mar  may  nov  oct  sep  
0         0  ...    0    0    0    0    0    0    1    0    0    0  
1         0  ...    0    0    0    0    0    0    1    0    0    0  
2         0  ...    0    0    0    0    0    0    1    0    0    0  
5         0  ...    0    0    0    0    0    0    1    0    0    0  
6         0  ...    0    0    0    0    0    0    1    0    0    0  

[5 rows x 40 columns]

In [17]:
df['age']= df['age']/df['age'].max()
df['balance']= df['balance']/df['balance'].max()
df['day']= df['day']/df['day'].max()
df['duration']= df['duration']/df['duration'].max()
df.head()

age  default   balance  housing  loan      day  duration  campaign  \
0  0.610526        0  0.020984        1     0  0.16129  0.053070         1   
1  0.463158        0  0.000284        1     0  0.16129  0.030704         1   
2  0.347368        0  0.000020        1     1  0.16129  0.015453         1   
5  0.368421        0  0.002262        1     0  0.16129  0.028264         1   
6  0.294737        0  0.004377        1     1  0.16129  0.044124         1   

   pdays  previous  ...  dec  feb  jan  jul  jun  mar  may  nov  oct  sep  
0     -1         0  ...    0    0    0    0    0    0    1    0    0    0  
1     -1         0  ...    0    0    0    0    0    0    1    0    0    0  
2     -1         0  ...    0    0    0    0    0    0    1    0    0    0  
5     -1         0  ...    0    0    0    0    0    0    1    0    0    0  
6     -1         0  ...    0    0    0    0    0    0    1    0    0    0  

[5 rows x 40 columns]

In [18]:
df['campaign']= df['campaign']/df['campaign'].max()
df['pdays']= df['pdays']/df['pdays'].max()
df['previous']= df['previous']/df['previous'].max()
df.head()

age  default   balance  housing  loan      day  duration  campaign  \
0  0.610526        0  0.020984        1     0  0.16129  0.053070  0.017241   
1  0.463158        0  0.000284        1     0  0.16129  0.030704  0.017241   
2  0.347368        0  0.000020        1     1  0.16129  0.015453  0.017241   
5  0.368421        0  0.002262        1     0  0.16129  0.028264  0.017241   
6  0.294737        0  0.004377        1     1  0.16129  0.044124  0.017241   

      pdays  previous  ...  dec  feb  jan  jul  jun  mar  may  nov  oct  sep  
0 -0.001148       0.0  ...    0    0    0    0    0    0    1    0    0    0  
1 -0.001148       0.0  ...    0    0    0    0    0    0    1    0    0    0  
2 -0.001148       0.0  ...    0    0    0    0    0    0    1    0    0    0  
5 -0.001148       0.0  ...    0    0    0    0    0    0    1    0    0    0  
6 -0.001148       0.0  ...    0    0    0    0    0    0    1    0    0    0  

[5 rows x 40 columns]

In [19]:
df.dtypes

age              float64
default            int64
balance          float64
housing            int64
loan               int64
day              float64
duration         float64
campaign         float64
pdays            float64
previous         float64
y                  int64
divorced           uint8
married            uint8
single             uint8
admin.             uint8
blue-collar        uint8
entrepreneur       uint8
housemaid          uint8
management         uint8
retired            uint8
self-employed      uint8
services           uint8
student            uint8
technician         uint8
unemployed         uint8
primary            uint8
secondary          uint8
tertiary           uint8
apr                uint8
aug                uint8
dec                uint8
feb                uint8
jan                uint8
jul                uint8
jun                uint8
mar                uint8
may                uint8
nov                uint8
oct                uint8
sep                uint8


Now lets check the correlation::

In [20]:
data1 = df[['y','age']]
correlation = data1.corr(method='pearson')
print(correlation)

           y      age
y    1.00000  0.02476
age  0.02476  1.00000


In [21]:
data1 = df[['y','default']]
correlation = data1.corr(method='pearson')
print(correlation)

                y   default
y        1.000000 -0.023244
default -0.023244  1.000000


In [22]:
data1 = df[['y','balance']]
correlation = data1.corr(method='pearson')
print(correlation)

                y   balance
y        1.000000  0.051386
balance  0.051386  1.000000


In [23]:
data1 = df[['y','housing']]
correlation = data1.corr(method='pearson')
print(correlation)

              y  housing
y        1.0000  -0.1383
housing -0.1383   1.0000


In [24]:
data1 = df[['y','loan']]
correlation = data1.corr(method='pearson')
print(correlation)

             y      loan
y     1.000000 -0.068802
loan -0.068802  1.000000


In [25]:
data1 = df[['y','day']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       day
y    1.000000 -0.030247
day -0.030247  1.000000


In [26]:
data1 = df[['y','duration']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  duration
y         1.000000  0.397393
duration  0.397393  1.000000


In [27]:
data1 = df[['y','campaign']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  campaign
y         1.000000 -0.072465
campaign -0.072465  1.000000


In [28]:
data1 = df[['y','pdays']]
correlation = data1.corr(method='pearson')
print(correlation)

              y     pdays
y      1.000000  0.101446
pdays  0.101446  1.000000


In [29]:
data1 = df[['y','previous']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  previous
y         1.000000  0.091764
previous  0.091764  1.000000


In [30]:
data1 = df[['y','divorced']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  divorced
y         1.000000  0.003044
divorced  0.003044  1.000000


In [31]:
data1 = df[['y','married']]
correlation = data1.corr(method='pearson')
print(correlation)

                y   married
y        1.000000 -0.060928
married -0.060928  1.000000


In [32]:
data1 = df[['y','single']]
correlation = data1.corr(method='pearson')
print(correlation)

               y    single
y       1.000000  0.064084
single  0.064084  1.000000


In [33]:
data1 = df[['y','admin.']]
correlation = data1.corr(method='pearson')
print(correlation)

               y    admin.
y       1.000000  0.007173
admin.  0.007173  1.000000


In [34]:
data1 = df[['y','blue-collar']]
correlation = data1.corr(method='pearson')
print(correlation)

                    y  blue-collar
y            1.000000    -0.070973
blue-collar -0.070973     1.000000


In [35]:
data1 = df[['y','entrepreneur']]
correlation = data1.corr(method='pearson')
print(correlation)

                     y  entrepreneur
y             1.000000     -0.019513
entrepreneur -0.019513      1.000000


In [36]:
data1 = df[['y','housemaid']]
correlation = data1.corr(method='pearson')
print(correlation)

                  y  housemaid
y          1.000000  -0.014936
housemaid -0.014936   1.000000


In [37]:
data1 = df[['y','management']]
correlation = data1.corr(method='pearson')
print(correlation)

                   y  management
y           1.000000    0.032032
management  0.032032    1.000000


In [38]:
data1 = df[['y','retired']]
correlation = data1.corr(method='pearson')
print(correlation)

                y   retired
y        1.000000  0.078686
retired  0.078686  1.000000


In [39]:
data1 = df[['y','self-employed']]
correlation = data1.corr(method='pearson')
print(correlation)

                      y  self-employed
y              1.000000       0.001161
self-employed  0.001161       1.000000


In [40]:
data1 = df[['y','services']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  services
y         1.000000 -0.028754
services -0.028754  1.000000


In [41]:
data1 = df[['y','student']]
correlation = data1.corr(method='pearson')
print(correlation)

                y   student
y        1.000000  0.073955
student  0.073955  1.000000


In [42]:
data1 = df[['y','technician']]
correlation = data1.corr(method='pearson')
print(correlation)

                 y  technician
y           1.0000     -0.0073
technician -0.0073      1.0000


In [43]:
data1 = df[['y','apr']]
correlation = data1.corr(method='pearson')
print(correlation)

           y      apr
y    1.00000  0.06292
apr  0.06292  1.00000


In [44]:
data1 = df[['y','aug']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       aug
y    1.000000 -0.008911
aug -0.008911  1.000000


In [45]:
data1 = df[['y','dec']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       dec
y    1.000000  0.074213
dec  0.074213  1.000000


In [46]:
data1 = df[['y','feb']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       feb
y    1.000000  0.037983
feb  0.037983  1.000000


In [47]:
data1 = df[['y','jan']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       jan
y    1.000000 -0.008908
jan -0.008908  1.000000


In [48]:
data1 = df[['y','jul']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       jul
y    1.000000 -0.036403
jul -0.036403  1.000000


In [49]:
data1 = df[['y','jun']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       jun
y    1.000000 -0.012643
jun -0.012643  1.000000


In [50]:
data1 = df[['y','mar']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       mar
y    1.000000  0.134693
mar  0.134693  1.000000


In [51]:
data1 = df[['y','may']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       may
y    1.000000 -0.101234
may -0.101234  1.000000


In [52]:
data1 = df[['y','nov']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       nov
y    1.000000 -0.014366
nov -0.014366  1.000000


In [53]:
data1 = df[['y','oct']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       oct
y    1.000000  0.124897
oct  0.124897  1.000000


In [54]:
data1 = df[['y','sep']]
correlation = data1.corr(method='pearson')
print(correlation)

            y       sep
y    1.000000  0.123879
sep  0.123879  1.000000


LETS MAKE THE MODEL

In [55]:
#classification

X= df[['duration','housing','pdays','campaign','previous','loan','balance','married','single','blue-collar','sep','oct','may','mar','dec','apr','student','retired']].values
X[0:5]

array([[ 5.30703538e-02,  1.00000000e+00, -1.14810563e-03,
         1.72413793e-02,  0.00000000e+00,  0.00000000e+00,
         2.09836772e-02,  1.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 3.07035380e-02,  1.00000000e+00, -1.14810563e-03,
         1.72413793e-02,  0.00000000e+00,  0.00000000e+00,
         2.83960167e-04,  0.00000000e+00,  1.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 1.54534364e-02,  1.00000000e+00, -1.14810563e-03,
         1.72413793e-02,  0.00000000e+00,  1.00000000e+00,
         1.95834598e-05,  1.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00,  0.00000000e+00,  0.00000000e

In [56]:
Y= df[['y']].values
Y[0:5]

array([[0],
       [0],
       [0],
       [0],
       [0]])

In [57]:
#normalize data
from sklearn import preprocessing
X= preprocessing.StandardScaler().fit(X).transform(X.astype(float))
X[0:5]

array([[ 0.01036799,  0.88208565, -0.41231121, -0.57382712, -0.25072965,
        -0.44378619,  0.25935402,  0.81530776, -0.6280861 , -0.52303563,
        -0.11167097, -0.1274133 ,  1.5080389 , -0.10237557, -0.06820499,
        -0.26428897, -0.13516858, -0.22859548],
       [-0.41572597,  0.88208565, -0.41231121, -0.57382712, -0.25072965,
        -0.44378619, -0.4355679 , -1.22653071,  1.59213839, -0.52303563,
        -0.11167097, -0.1274133 ,  1.5080389 , -0.10237557, -0.06820499,
        -0.26428897, -0.13516858, -0.22859548],
       [-0.70624459,  0.88208565, -0.41231121, -0.57382712, -0.25072965,
         2.25333735, -0.44444344,  0.81530776, -0.6280861 , -0.52303563,
        -0.11167097, -0.1274133 ,  1.5080389 , -0.10237557, -0.06820499,
        -0.26428897, -0.13516858, -0.22859548],
       [-0.46220895,  0.88208565, -0.41231121, -0.57382712, -0.25072965,
        -0.44378619, -0.36916571,  0.81530776, -0.6280861 , -0.52303563,
        -0.11167097, -0.1274133 ,  1.5080389 , -0.102

In [58]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.8,random_state=4)
print('train set: ',X_train.shape, Y_train.shape)
print('test set: ',X_test.shape, Y_test.shape)

train set:  (8638, 18) (8638, 1)
test set:  (34555, 18) (34555, 1)


In [59]:
#now for classification
from sklearn.neighbors import KNeighborsClassifier

In [95]:
#training
k=9
neigh = KNeighborsClassifier(n_neighbors=k).fit(X_train,Y_train)
neigh

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=9, p=2,
           weights='uniform')

In [96]:
#predicting
yhat= neigh.predict(X_test)
yhat[0:5]

array([0, 1, 0, 0, 0])

In [97]:
#accuracy evaluation
from sklearn import metrics
print("Train set accuracy: ", metrics.accuracy_score(Y_train,neigh.predict(X_train)))
print("Test set accuracy: ", metrics.accuracy_score(Y_test,yhat))

Train set accuracy:  0.9076175040518638
Test set accuracy:  0.8900303863406164


Thus we get test accuracy of about 89% on 80% test data